src/feedback/fir.rs

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::collections::HashMap;
+
+use crate::feedback::FciFeedbackPacketType;
+use crate::prelude::*;
+use crate::utils::u32_from_be_bytes;
+use crate::{RtcpParseError, RtcpWriteError};
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct FirEntry {
+    ssrc: u32,
+    sequence: u8,
+}
+
+impl FirEntry {
+    fn new(ssrc: u32, sequence: u8) -> Self {
+        Self { ssrc, sequence }
+    }
+
+    pub fn ssrc(&self) -> u32 {
+        self.ssrc
+    }
+
+    pub fn sequence(&self) -> u8 {
+        self.sequence
+    }
+}
+
+pub struct FirParserEntryIter<'a> {
+    parser: &'a Fir<'a>,
+    i: usize,
+}
+
+impl<'a> FirParserEntryIter<'a> {
+    fn decode_entry(entry: &[u8]) -> FirEntry {
+        FirEntry::new(u32_from_be_bytes(&entry[0..4]), entry[4])
+    }
+}
+
+impl<'a> std::iter::Iterator for FirParserEntryIter<'a> {
+    type Item = FirEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let idx = self.i * 8;
+        if idx + 7 >= self.parser.data.len() {
+            return None;
+        }
+        let entry = FirParserEntryIter::decode_entry(&self.parser.data[idx..]);
+        self.i += 1;
+        Some(entry)
+    }
+}
+
+/// FIR (Full Intra Refresh) information as specified in RFC 5104
+pub struct Fir<'a> {
+    data: &'a [u8],
+}
+
+impl<'a> Fir<'a> {
+    /// The list of RTP SSRCs that are requesting a Full Intra Refresh.
+    pub fn entries(&self) -> impl Iterator<Item = FirEntry> + '_ {
+        FirParserEntryIter { parser: self, i: 0 }
+    }
+
+    pub fn builder() -> FirBuilder {
+        FirBuilder::default()
+    }
+}
+
+impl<'a> FciParser<'a> for Fir<'a> {
+    const PACKET_TYPE: FciFeedbackPacketType = FciFeedbackPacketType::PAYLOAD;
+    const FCI_FORMAT: u8 = 4;
+
+    fn parse(data: &'a [u8]) -> Result<Self, RtcpParseError> {
+        if data.len() < 8 {
+            return Err(RtcpParseError::Truncated {
+                expected: 8,
+                actual: data.len(),
+            });
+        }
+        Ok(Self { data })
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct FirBuilder {
+    ssrc_seq: HashMap<u32, u8>,
+}
+
+impl FirBuilder {
+    pub fn add_ssrc(mut self, ssrc: u32, sequence: u8) -> Self {
+        self.ssrc_seq
+            .entry(ssrc)
+            .and_modify(|entry| {
+                *entry = sequence;
+            })
+            .or_insert(sequence);
+        self
+    }
+}
+
+impl<'a> FciBuilder<'a> for FirBuilder {
```

Comment:
Can be `'static` here since `FirBuilder` always owns its field.

```rust
impl FciBuilder<'static> for FirBuilder {
    [...]
}
```

Also applicable for most other feedback packet builders, except `RpsiBuilder` afaict. Forgot to mention it in previous PR.

Reply:
Hmm... Not so sure about this now. Let me think about it more. I'm still trying to have the `TransportFeedbackBuilder` & `PayloadFeedbackBuilder` handle the `fci` as either borrowed or owned and it seems like keeping the `'a` bound would help there.